source/connections-survive-step-down/tests/README.rst

```diff
@@ -0,0 +1,224 @@
+"server shutting down" error responses from the server.
+
+.. contents::
+
+These tests apply only to replica set topologies.
+
+Tests
+-----
+
+The driver should implement the following tests:
+
+getMore Iteration
+`````````````````
+
+This test requires server version 4.2 or higher, and makes use of events
+defined in the `CMAP specification
+<https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst>`_.
+
+Perform the following operations:
+
+- Insert 5 documents into a collection.
+- Clear the connection pool to the current primary (see CMAP spec for
+  the definition of connection pool clear operation, and
+  `Why clear the pool in getMore test?`_ for rationale).
+- Start a find operation on the collection with a batch size of 2, and
+  retrieve the first batch of results.
+- Verify that a ConnectionCreated CMAP event has been published.
+- Send a ``replSetStepDown`` command to the current primary and verify that
+  the command succeeded, or follow `Server Step Down Procedure`_ below.
+- Retrieve the next batch of results from the cursor obtained in the find
+  operation, and verify that this operation succeeded.
+- Verify that no new ConnectionCreated CMAP events have been published.
+
+Not Master - Keep Connection Pool
+`````````````````````````````````
+
+This test requires server version 4.2 or higher.
+
+Perform the following operations:
+
+- Perform server selection for the primary server to obtain a server object
+  with an initialized connection pool.
+- Mock an operation returning NotMaster operation failure error code.
+- Verify that:
+
+  - Connection pool generation (per CMAP spec) has not changed
+  - PoolCleared CMAP event has not been published
+
+Not Master - Reset Connection Pool
+``````````````````````````````````
+
+This test requires server version 4.0 or lower.
+
+Perform the following operations:
+
+- Perform server selection for the primary server to obtain a server object
+  with an initialized connection pool.
+- Mock an operation returning NotMaster operation failure error code.
+- Verify that:
+  - Connection pool generation (per CMAP spec) has been incremented by 1
+  - PoolCleared CMAP event has been published
+
+Shutdown - Reset Connection Pool
+````````````````````````````````
+
+This test should be run on all supported server versions.
+
+Perform the following operations:
+
+- Perform server selection for the primary server to obtain a server object
+  with an initialized connection pool.
+- Mock an operation returning ShutdownInProgress (91) operation
+  failure error code.
+- Verify that:
+  - Connection pool generation (per CMAP spec) has been incremented by 1
+  - PoolCleared CMAP event has been published
+
+
+Server Step Down Procedure
+--------------------------
+
+In principle, the current primary may be stepped down by running the following
+admin command on it: ``{replSetStepDown: null}``. This will cause the current
```

Comment:
Do you need force: true?

Reply:
That would probably speed things up and avoid the math formula pertaining to `secondaryCatchUpPeriodSecs`. I don't think this test needs to care about rollbacks, so we can probably just use `force: true` to run as quickly as possible.

Reply:
`force: true` was not part of the guidance I received from the server team with respect to step downs.

Reply:
If there is a rollback, since we are reading data that was just inserted, would it not be possible for the read to fail?

Reply:
Not if the inserts are done with `w:majority`.

Reply:
If it will speed things up, let's do it.

Reply:
I added `force: true` to the simple command, however a forced step down still takes much longer (over 10 seconds in my testing) than one performed using the extended method.

Both seem to work, thus a driver may choose to implement the forced or the extended method depending on performance characteristics they would like their test suite to have (i.e. whether it is OK to wait 15 seconds for an election).